### PR TITLE
[NEW] Allow the user to update avatar from profile view

### LINKED ISSUE #752: [NEW] Update user avatar
From the profile view, allow to the user change its avatar picture. The idea is simple: have a action to open the [Storage Access Framework](https://developer.android.com/guide/topics/providers/document-provider.html) (only - obvious - for images) and call the [client.setAvatar(file: File, mimeType: String)](https://github.com/RocketChat/Rocket.Chat.Kotlin.SDK/blob/99b4ce6c2e79a7c6f7a4d2a9e3a8b2b719b94720/core/src/main/kotlin/chat/rocket/core/internal/rest/User.kt#L93)

Tip: see the existing upload file feature.

### PR DESCRIPTION
**Pull request by aniketsingh03**
@RocketChat/android

Closes #752 
 
**DESCRIPTION**
The user will be able to choose avatar either by clicking an image from camera or by choosing an image from storage.

### CODE REVIEW

**Comment by luciofm** [comment]
@aniketsingh03 will take a look later today, thanks for the PR.

**Comment by rafaelks** [comment]
@aniketsingh03 Is this PR ready to be reviewed already?

**Comment by aniketsingh03** [comment]
Yeah @rafaelks , you may review this, however this depends on [#77](https://github.com/RocketChat/Rocket.Chat.Kotlin.SDK/pull/77) and [#71](https://github.com/RocketChat/Rocket.Chat.Kotlin.SDK/pull/71) of the sdk as it is blocked by [#762](https://github.com/RocketChat/Rocket.Chat.Android/issues/762). You may merge this once you are sure to have resolved those.

**Comment by leonardoaramaki** [comment]
@aniketsingh03 Looks like the payload we're sending is wrong. We should send data.email, data.name and data.password.

**Comment by aniketsingh03** [comment]
@leonardoaramaki are you talking about the payload for `updateProfile` ?

**Comment by leonardoaramaki** [comment]
@aniketsingh03 :+1:

**Comment by aniketsingh03** [comment]
You might want to have a look at https://github.com/RocketChat/Rocket.Chat.Android/issues/762 @leonardoaramaki .

**Comment by aniketsingh03** [comment]
@filipedelimabrito this pr is ready to be merged. I have resolved the conflicts too.

**Comment by aniketsingh03** [comment]
@rafaelks can you have a look at this one?

**Comment by aniketsingh03** [comment]
@filipedelimabrito @leonardoaramaki @rafaelks ^

File: app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt]: Could you fix this wrong indentation? I know that when using `.ifNull {` the IDE is putting the wrong indentation, but we should indent this correct (in this case aligned with `view.showMessage(it)`)

File: app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt]: Here too.

File: app/src/main/res/values/strings.xml
Comment (philipbrito) [inline app/src/main/res/values/strings.xml:144]: Could you add to the strings.xml in `values-pt-rBR` the following value:

`<string name="permission_image_picking_not_allowed">A imagem não pode ser acessada sem a permissão para ler o armazenamento.</string>`

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: We aren't using the Hungarian notation on the project (code style). Could you change it?

File: app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt]: First you are [allowing to the user the selection of any image type](https://github.com/aniketsingh03/Rocket.Chat.Android/blob/b811400eec76aa399ccba84039f596fe02b6d9d1/app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt#L153) but here you are specifying that any image picked by the user is a JPEG image. 
We added an [extension function](https://github.com/aniketsingh03/Rocket.Chat.Android/blob/b811400eec76aa399ccba84039f596fe02b6d9d1/app/src/main/java/chat/rocket/android/util/extensions/Uri.kt#L44) to get the image type from an URI: 

 `yourUri.getMimeType(aContext)`

Could you change it?

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: context!!

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: activity!!

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: You need to add a string entry in the strings.xml file.
Also, change `Choose photo from storage` to `Choose photo from gallery`.

Translations for PT-BR: `Tirar foto`. `Escolher foto da galeria`.

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: data!!

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (aniketsingh03) [reply app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: What changes need to be made here @filipedelimabrito ?

**Comment by aniketsingh03** [comment]
@filipedelimabrito I have made the suggested changes

**Comment by aniketsingh03** [comment]
@filipedelimabrito @luciofm please tell me if any more changes are required.

**Comment by aniketsingh03** [comment]
@filipedelimabrito @rafaelks ^^

**Comment by aniketsingh03** [comment]
@leonardoaramaki can you have a look at this ?

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: You need to check if the data intent isn't null after the `resultCode == RESULT_OK`

```
if (resultCode == RESULT_OK) {
 if (data != null) {
  [...]
 }
}
```

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: Add curly brackets here.

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: Why are you adding another ` Observables.combineLatest(..)` here? You can user the other one.

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/res/values/strings.xml
Comment (aniketsingh03) [reply app/src/main/res/values/strings.xml]: This is already added.

**Comment by aniketsingh03** [comment]
@filipedelimabrito I have dealt with all edge cases, like the case when the user enters both the avatar url and chooses an image from gallery, then an alert dialog will be displayed to notify the user to choose either of them. I do not have the translations for the strings added in the latest commits, can you provide me translations for the same?

**Comment by aniketsingh03** [comment]
@filipedelimabrito ^^

File: app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomFragment.kt
Comment (leonardoaramaki) [inline app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomFragment.kt]: By doing this you're hiding the toolbar, which is not exactly what we're looking for. :(

**Review by leonardoaramaki — CHANGES_REQUESTED** [review]
Thanks for your PR! I made a comment on the keyboard resize part of the code. Also, testing here on my device after I took a picture for my avatar the avatar remained blank and when I tapped on the done check mark the app crashed. 
Here's the line where it crashed: https://github.com/RocketChat/Rocket.Chat.Android/blob/fdaedd8520baa9a7378202d52f818b9b7f011ac2/app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt#L62-L62

**Comment by aniketsingh03** [comment]
@leonardoaramaki can you provide a complete log so that I may have a look as to why the app is crashing?

**Comment by aniketsingh03** [comment]
@leonardoaramaki ^

**Comment by aniketsingh03** [comment]
@leonardoaramaki @filipedelimabrito I have fixed the keyboard issue, please have a look and suggest any changes to be made. Also @leonardoaramaki can you provide me the logs for the crash so that I may try to fix the other issue too?

File: app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt
Comment (leonardoaramaki) [inline app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt]: Just a minor point. It's more idiomatic on Kotlin if instead you check like `if (avatarUrl.isNotEmpty()) {`.

File: app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomFragment.kt
Comment (leonardoaramaki) [inline app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomFragment.kt]: Can you please remove this? We've already solved this keyboard issue on another PR. :)

**Review by leonardoaramaki — CHANGES_REQUESTED** [review]

**Comment by aniketsingh03** [comment]
@leonardoaramaki I have made the requested changes. Can you provide the translations for the newly added strings? Some tests are failing because of lint errors which are probably occurring in all pull requests, so I guess we can have a separate pr which fixes all the lint errors.

**Comment by aniketsingh03** [comment]
@leonardoaramaki @filipedelimabrito can this be merged?

**Comment by rafaelks** [comment]
@aniketsingh03 I'm sorry for this, but your PR has some new conflicts. Can you solve them please?

**Comment by rafaelks** [comment]
@aniketsingh03 pt-BR strings are looking good! 👍

**Comment by aniketsingh03** [comment]
@rafaelks @filipedelimabrito all changes have been made, you may merge this.

**Comment by rafaelks** [comment]
@aniketsingh03 App is crashing for me when opening the Profile view. I'm on Android 6.0, can you check please?

**Comment by aniketsingh03** [comment]
@rafaelks can you please provide a log of the error obtained when the app crashed?

**Comment by aniketsingh03** [comment]
@rafaelks resolved some conflicts, have a look.

**Comment by aniketsingh03** [comment]
@rafaelks I have resolved all the conflicts, please have a look.

**Comment by philipbrito** [comment]
@aniketsingh03 Could you resolve the conflicts more one time? Thanks.

File: app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/presentation/ProfilePresenter.kt:62]: 👍

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: `text_avatar_url.textContent.isNotBlank()`

File: app/src/main/res/values/strings.xml
Comment (philipbrito) [inline app/src/main/res/values/strings.xml:18]: What about adding these strings to our Hindi file too?

Comment (aniketsingh03) [reply app/src/main/res/values/strings.xml:18]: Cool, I'll add them :+1:

**Comment by aniketsingh03** [comment]
@filipedelimabrito I have made the suggested changes.

File: app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/profile/ui/ProfileFragment.kt]: Could you change it to `text_avatar_url.isNotBlank()`?
Thanks.

**Review by philipbrito — CHANGES_REQUESTED** [review]

**Review by philipbrito — CHANGES_REQUESTED** [review]
Code is ok now but I tested how the app works when updating the avatar and here are some points that needs to be addressed in this PR:

- When giving permission to access the phone storage shows to the user the `alert dialog` immediately (currently we have to tap two times to change the avatar from the app for the first time: tap to change the avatar > give the permission > tap to change the avatar again).

- When the user chose to not give the app permission to access the storage, show to the user a message that in order to change the avatar we need that permission (you could use the `toast` to do so).

- Ok, I selected/took a photo and decided to not upload the avatar (tapping the back button on the action mode), what happens? It stills shows the photo picked from the gallery or took from the camera. What I expected? The app should show my current avatar.

- Also, this is **critical**: Try uploading several avatars in sequence, the profile screen shows an avatar uploaded in another time (not the current avatar).

**Comment by aniketsingh03** [comment]
@filipedelimabrito I have tried to cover all the edge cases but I might have missed one, have a look and suggest changes. As far as the 4th point is concerned, I feel that this is an issue with `SimpleDraweeView`, can't we have a separate issue for that?

**Comment by aniketsingh03** [comment]
@filipedelimabrito @rafaelks can this be merged?

**Comment by rafaelks** [comment]
@aniketsingh03 Can you please fix the conflicts on this one?

**Comment by aniketsingh03** [comment]
@filipedelimabrito please have a look at this one.

**Comment by rafaelks** [comment]
This one is going to be resolved by #1527.

**Pull request closed without merging**
